### PR TITLE
Rework the use of completable futures

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

**Comment by henryju** [comment]
@vojtech-suchy-sonarsource sorry, this is a big PR, but we have hopefully already discussed many of the changes.

File: rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4j.jsonrpc.CompletableFutures;
+import org.springframework.beans.factory.BeanFactory;
+
+class AbstractRpcServiceDelegate {
```
Comment (vojtech-suchy-sonarsource) [inline rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java:32]: Should we mark this class `abstract`?

File: rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      var wrapper = new CancelCheckerWrapper(cancelChecker);
+      wrapper.checkCanceled();
+      return wrapper;
+    }).thenApplyAsync(cancelChecker -> {
+      cancelChecker.checkCanceled();
+      return code.apply(cancelChecker);
+    }, requestsExecutor);
+  }
+
+  protected <R> CompletableFuture<R> runAsync(Consumer<CancelChecker> code) {
```
Comment (vojtech-suchy-sonarsource) [inline rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java:59]: We could lose the generic type and just use `CompletableFuture<Void>` 
(The callers also have `CompletableFuture<Void>` as a return type)

Comment (henryju) [reply rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java:59]: Good catch, that was the point of this method :)

File: rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java (hunk 1 above)
Comment (henryju) [reply rpc-impl/src/main/java/org/sonarsource/sonarlint/core/rpc/impl/AbstractRpcServiceDelegate.java:32]: yes

**Review by vojtech-suchy-sonarsource — APPROVED** [review]

**Pull request merged**